File: model/mqtt_access.py
import sys
import os
import json
import paho.mqtt.client as mqtt

sys.path.append(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

# MQTT broker details
MQTT_BROKER = "150.140.186.118"
MQTT_PORT = 1883
MQTT_TOPIC = "v3_fanaria/#"  # Subscribe to all subtopics under v2_fanaria


# Callback when the client receives a CONNACK response from the server
def on_connect(client, userdata, flags, rc):
    print(f"Connected with result code {rc}")
    # Subscribing to the topic
    client.subscribe(MQTT_TOPIC)


# Callback when a PUBLISH message is received from the server
def on_message(client, userdata, msg):
    global message_received
    try:
        payload = json.loads(msg.payload.decode())
        pretty_payload = json.dumps(payload, indent=4)
        print(f"Topic: {msg.topic}\nMessage: {pretty_payload}")
        message_received = True
        print("\nMessage received=", message_received)
    except json.JSONDecodeError:
        print(f"Topic: {msg.topic}\nMessage: {msg.payload.decode()}")
        message_received = True
        print("\nMessage received=", message_received)


# Create an MQTT client instance
client = mqtt.Client()

# Assign the callback functions
client.on_connect = on_connect
client.on_message = on_message

# Connect to the MQTT broker
client.connect(MQTT_BROKER, MQTT_PORT, 60)

# Start the MQTT client loop to process network traffic and dispatch callbacks
client.loop_forever()

File: public/mqtt_access.mjs
import mqtt from "mqtt";

// MQTT broker details
const MQTT_BROKER = "mqtt://150.140.186.118:1883";
let currentTopic = "v2_fanaria/v2_omada14_fanari_0"; // Initial topic to subscribe to

// Create an MQTT client instance
const client = mqtt.connect(MQTT_BROKER);

// Callback when the client connects to the broker
client.on("connect", () => {
  console.log("Connected to MQTT broker");
  // Subscribe to the initial topic
  subscribeToTopic(currentTopic);
});

// Function to subscribe to a topic
export function subscribeToTopic(topic) {
  client.subscribe(topic, (err) => {
    if (!err) {
      console.log(`Subscribed to topic: ${topic}`);
    } else {
      console.error(`Failed to subscribe to topic: ${topic}`, err);
    }
  });
}

// Function to unsubscribe from a topic
export function unsubscribeFromTopic(topic) {
  client.unsubscribe(topic, (err) => {
    if (!err) {
      console.log(`Unsubscribed from topic: ${topic}`);
    } else {
      console.error(`Failed to unsubscribe from topic: ${topic}`, err);
    }
  });
}

// Callback when a message is received
client.on("message", (topic, message) => {
  try {
    // const payload = JSON.parse(message.toString());
    console.log(
      `Topic: ${topic}\nMessage: There is a change in topic ${topic}`
    );
  } catch (error) {
    console.error(`Failed to parse message on topic: ${topic}`, error);
  }
});

// Example function to change subscription to a new fanari
function changeSubscription(newFanariId) {
  const newTopic = `v2_fanaria/v2_omada14_fanari_${newFanariId}`;
  unsubscribeFromTopic(currentTopic);
  currentTopic = newTopic;
  subscribeToTopic(currentTopic);
}

// Example usage: Change subscription to fanari 1 after 10 seconds
setTimeout(() => {
  changeSubscription(1);
}, 10000);
